### PR TITLE
doc: adding Database Design document

### PR DESCRIPTION
**Pull request by sadiqkhoja**
Created this to document our database design. It should help in onboarding new contributors.

I have created ER Diagrams using Mermaid.js and added some description about them. I think we should add some flow diagrams in Contributing.md as well.

Feel free to make changes to the document as part of the review.

### CODE REVIEW

**Comment by alxndrsn** [comment]
You can see what the diagrams look like at https://github.com/sadiqkhoja/central-backend/blob/docs/erd/docs/database.md

File: docs/database.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+- A User is an Actor.
+- A Public Link (public_links) is an Actor, it is generated for anonymous submission through Enketo (web).
+- A Field Key (field_keys) is an Actor, it is generated for an ODK Collect App User.
+- An Actor can create Entities.
+- An Actor can have HTTP Sessions (sessions). We keep session IDs in the database to support distributed application instances.
+- An Actor can make Comments on Submissions.
+- An Actor can create Submissions and its Definitions (submission_def).
+- An Actor is assigned a Role on an Actee. It is a three way many-to-many relationship, resolved in assignments table.
+- An Actor can perform actions on the Actees, the action is logged in Audits (audits) table
+- Actees table contains autogenerated UUID for different species (entities as in Entitiy-Relationship) of the database like Projects, Forms, Users, etc. This is a clever way to handle access management that allow us to know the permissions of an user without having to join all the tables to Assignments table. Additionally we keep audit logs using actee ID, so that even if original object of that actee is deleted we can trace the history.
```
Comment (ktuite) [inline docs/database.md:145]: "Entitiy" -> "Entity"

**Review by ktuite — COMMENTED** [review]
This looks great! I really like seeing the DB laid out like this and explained. 

I'm thinking about how we could link this from the contribution guide and how it would be good to mention in

1. The "Framework Concepts" section about database logic/organization/validation
2. Probably also in "Navigating the Code", which needs some updating as it has references to things like `model/trait` that aren't there anymore.

File: docs/database.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+ >_(The diagram has actual table names in the boxes but in the following description, I have used entity nouns and put table name in the brackets where necessary)_
+
+Forms and Submissions are the core part of ODK Central. Above ER Diagram shows how we store Forms, Submissions and everything that is related to them in the database. 
+
+
+- A Project can have many Forms.
+- A Form is an Actee, which mean some action can be performed on it.
+- Each Form has one or more Definitions (form_defs) to support multiple versions. We keep form definition in  XForm (XML) format in a single column of form_defs table. If XLSForm was uploaded then .XLS file is kept in Blobs (blobs) table.
+- Each Form can have Fields (form_fields) i.e. questions, notes, etc.
```
[the remark this answers is not in the document]
Comment (sadiqkhoja) [reply docs/database.md:55]: Minimum cardinality is 0 that's why "can have"

File: docs/database.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    form_attachments }o..o| blobs : ""
+
+```
+
+ >_(The diagram has actual table names in the boxes but in the following description, I have used entity nouns and put table name in the brackets where necessary)_
+
+Forms and Submissions are the core part of ODK Central. Above ER Diagram shows how we store Forms, Submissions and everything that is related to them in the database. 
+
+
+- A Project can have many Forms.
```
[the remark this answers is not in the document]
Comment (sadiqkhoja) [reply docs/database.md:55]: In this section, I just want to focus on Project, Forms and Submissions. I have added description about Actor in the third section.

File: docs/database.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+```
+ 
+ >_(The diagram has actual table names in the boxes but in the following description, I have used entity nouns and put table name in the brackets where necessary)_
+
+This area of the database deals with the users, their access rights and audit logs.
+
+- A User is an Actor.
+- A Public Link (public_links) is an Actor, it is generated for anonymous submission through Enketo (web).
+- A Field Key (field_keys) is an Actor, it is generated for an ODK Collect App User.
+- An Actor can create Entities.
```
[the remark this answers is not in the document]
Comment (sadiqkhoja) [reply docs/database.md:139]: From DB perspective, we don't have createdBy column in Projects table, hence no foreign-key constraint. Whereas, in Entities table we have a direct link with Actors table

**Review by ktuite — APPROVED** [review]

**Pull request merged**